src/helpers.ts

```diff
@@ -1,243 +1,148 @@
+  environment: string,
+  first: number = 20,
+  cursor?: string
+): Promise<Deployments> {
+  // Query the deployments in the environment
+  // Sort by createdAt descending, so the most recent deployment is first
+  // Limit the number of deployments to the last 20
+  // If a cursor is provided, use that to paginate to the next page of results
+  const data = await octokit.graphql<DeploymentsGraphQLResponse>(
+    `
+    query fetchDeployments($owner: String!, $repo: String!, $environment: String!, $first: Int!, $cursor: String) {
+      repository(owner: $owner, name: $repo) {
+        deployments(
+          first: $first,
+          environments: [$environment],
+          orderBy: {field: CREATED_AT, direction: DESC},
+          after: $cursor
+        ) {
+          nodes {
+            databaseId
+            state
+          }
+          pageInfo {
+            hasNextPage
+            endCursor
+          }
         }
       }
     }
+  `,
+    {
+      first,
+      cursor,
+      environment,
+      owner: context.owner,
+      repo: context.repo
+    }
   )
 
-  if (result !== 0) {
-    throw new Error('Failed to get git log')
-  }
-
-  const commits = shas.split('\n').map(commit => {
-    // Split sha and message
-    const [sha, ...message] = commit.split(' ')
-    return { sha, message: message.join(' ') } // Not very efficient, but it's a small string
-  })
-
-  return commits
+  // Return the deployments
+  return data.fetchDeployments.repository.deployments
 }
 
-export async function releaseSha(
+/**
+ * Finds the most recent active deployment in specified environment, if any.
+ * Politely pages through the deployments using the GitHub GraphQL API.
+ * We return the databaseId of the deployment, which can be used to fetch the deployment
+ * from the GitHub REST API.
+ * @param octokit
+ * @param context
+ * @param environment
+ * @returns the deployment id if found, otherwise null
+ */
+export async function fetchDeploymentStatus(
   octokit: Octokit,
   context: Context,
-  environment: string
-): Promise<string> {
-  // Get deployment statuses for the selected environment
-  const deploymentStatuses = await octokit.rest.repos.listDeployments({
-    owner: context.repo.owner,
-    repo: context.repo.repo,
-    environment
-  })
-
-  let previousSha: string
-
-  // Find commit ref for the latest deployment, if any
-  if (deploymentStatuses.data.length > 0) {
-    const latestDeployment = deploymentStatuses.data[0]
-    previousSha = latestDeployment.sha
-  } else {
-    previousSha = context.sha
-  }
-
-  return previousSha
-}
-
-export async function processCommits(
-  commits: GitLog[],
-  workspace: string
-): Promise<GitLog[]> {
-  const relevantCommits: GitLog[] = []
-
-  // Checkout commit using shell script
-  for (const commit of commits) {
-    const checkout = await exec('git', ['checkout', commit.sha])
-    if (checkout !== 0) {
-      continue
+  environment: string,
+  nth: number
+): Promise<number | null> {
```

Comment:
For the future, is there a way to cache this? If we see a previously checked commit and still no new deployments, just returned the previous result?

```suggestion
): Promise<number | null> {
  // TODO: Return previous result if we find a commit that's been checked previously
  // and there are still no new commits
```

Reply:
Deployments and releases can be deleted by users, so re-running is the safest bet. Optional caching could perhaps be something to look into. Normally it doesn't have to page that far.